### ISSUE TITLE
Remove support for the catch all (*) in the using statement

### ISSUE THREAD

**Issue by axic** [ethereum/solidity#11882]
Does the `using MyLIb for *;` statement saves that much typing/space as opposed to clearly typing out the intended types?

**Comment by cameel** [comment]
There are some types you cannot name but for which you can add still functions with `*`. Most notably literals:

```solidity
library Math {
    function add(uint a, uint b) internal returns (uint) {
        return a + b;
    }
}

contract C {
    using Math for uint;
    function f() public returns (uint) { return 2.add(2); } // Error: Member "add" not found or not visible after argument-dependent lookup in int_const 2.
}

contract D {
    using Math for *;
    function f() public returns (uint) { return 2.add(2); } // OK
}
```

**Comment by axic** [comment]
I think that should be solved in a nicer way (i.e. having a rationalnumber type), than this abuse :)

**Comment by cameel** [comment]
Agreed. I think that `*` is a bad practice in general. Just wanted to point out we probably cannot remove it without providing a replacement for this. Unless it turns out that no one knows about this trick and therefore no one uses it :)

**Comment by axic** [comment]
> Unless it turns out that no one knows about this trick and therefore no one uses it :)

Now they certainly know about it 😅

**Comment by chriseth** [comment]
I'm fine with requiring to be explicit, but maybe we should ask if this breaks anything.

**Comment by chriseth** [comment]
Oh and we should allow `... for {uint, string};` in this case, i.e. providing a list of types.

**Comment by axic** [comment]
> Oh and we should allow `... for {uint, string};` in this case, i.e. providing a list of types.

That probably is a good idea.

**Comment by chriseth** [comment]
We should/could check if implicit conversions are still possible on the self type as long as all types are explicitly listed in the using statment, i.e. `using L.add for {uint8, uint16, ...};` allows a single implementation of the SafeMath `add` function.

**Comment by cameel** [comment]
On today's design call we decided to postpone the deprecation of this syntax until we actually have a replacement for all of its functionality. It will not be deprecated in 0.9.0, but the overall plan is still to deprecate and remove it eventually.
